### PR TITLE
Keep global step update in the loop

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

The gradient accumulation logic check and global step update should be contained within the loops.
This PR changes the TTP API so that we only use it to know whether the optimization is handled internally in the case of the IPU and DeepSpeed plugins.

### Does your PR introduce any breaking changes? If yes, please list them.

Beta `TrainingTypePlugin` and `Accelerator` API:

- Replaced `update_global_step` with the `accumulate_grad_batches` property.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 1]
```diff
@@ -248,7 +249,8 @@ def _run_validation(self):
 
     def _accumulated_batches_reached(self) -> bool:
         """Determine if accumulation will be finished by the end of the current batch."""
-        return self.batch_progress.current.ready % self.trainer.accumulate_grad_batches == 0
+        agb = self.trainer.accelerator.accumulate_grad_batches or self.trainer.accumulate_grad_batches
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:252]: do you know, are we handling the case when both are set elsewhere?

Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:252]: When the accelerator property is not `None`, the trainer property will also be set.

But we want the accelerator to take precedence in that case.

Is that what you were asking?

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:252]: yeah my bad, I thought trainer.accumulate_grad_batches is None by default but it is 1, so we wouldn't actually error if both are set and instead document the precedence.

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 2]
```diff
@@ -180,9 +180,8 @@ def _run_optimization(
         result = AttributeDict()
         closure = self._make_closure(split_batch, batch_idx, opt_idx, optimizer, self._hiddens, result)
 
-        if self.trainer.fit_loop.should_accumulate():
-            # For gradient accumulation
-
+        # when the training type plugin handles accumulation, we want to always call the optimizer step
+        if not self.trainer.accelerator.handles_accumulate_grad_batches and self.trainer.fit_loop.should_accumulate():
```
Comment (tchaton) [inline pytorch_lightning/loops/batch/training_batch_loop.py:184]: Can we move this logic under `should_accumulate()`. For loop customization, users should not have to know training type plugin handles accumulation differently.

Comment (carmocca) [reply pytorch_lightning/loops/batch/training_batch_loop.py]: We can't because we need to call the `optimizer.step` if the TTP handles it as they also track whether accumulation should be applied.

File: tests/plugins/test_deepspeed_plugin.py [hunk 3]
```diff
@@ -717,14 +717,19 @@ def on_train_batch_start(
     trainer = Trainer(
         default_root_dir=tmpdir,
         progress_bar_refresh_rate=0,
-        max_epochs=5,
+        # TODO: this test fails with max_epochs >1 as there are leftover batches per epoch.
+        # there's divergence in how Lightning handles the last batch of the epoch with how DeepSpeed does it.
+        # we step the optimizers on the last batch but DeepSpeed keeps the accumulation for the next epoch
```
Comment (carmocca) [inline tests/plugins/test_deepspeed_plugin.py:722]: This is also a problem in master with an odd `limit_train_batches` and `max_epochs=2`

`$ PL_RUNNING_SPECIAL_TESTS=1 pytest -v --capture=no --disable-warnings tests/plugins/test_deepspeed_plugin.py::test_deepspeed_multigpu_stage_2_accumulated_grad_batches`

**Review by tchaton — APPROVED** [review]
LGTM ! Nice cleanup.

File: tests/plugins/test_deepspeed_plugin.py (hunk 3 above)
Comment (tchaton) [reply tests/plugins/test_deepspeed_plugin.py:722]: Any ideas ? I am not sure to grasp the problem.

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 4]
```diff
@@ -517,18 +521,10 @@ def optimizer_step(self, optimizer: torch.optim.Optimizer, lambda_closure: Calla
...
-        accumulation logic internally.
-        """
-        if self.config.get("gradient_accumulation_steps") > 1:
-            self._original_accumulate_grad_batches = self.lightning_module.trainer.accumulate_grad_batches
-            # todo (tchaton) Add support for accumulate_grad_batches being a dictionary.
-            self.lightning_module.trainer.accumulation_scheduler = GradientAccumulationScheduler({0: 1})
-        else:
-            self._original_accumulate_grad_batches = None
+    @property
+    def handles_accumulate_grad_batches(self) -> bool:
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/deepspeed.py:525]: "accumulate_grad_batches" was not a great name from the beginning imo. do we want to take this opportunity and choose the right name? like `handles_gradient_accumulation` for example?

File: tests/accelerators/test_ipu.py [hunk 5]
```diff
@@ -312,7 +312,7 @@ def test_accumulate_grad_batches_dict_fails(tmpdir):
     model = IPUModel()
     trainer = Trainer(default_root_dir=tmpdir, ipus=1, accumulate_grad_batches={0: 1})
     with pytest.raises(
-        MisconfigurationException, match="IPUs currently only support accumulate_grad_batches being an integer value."
+        MisconfigurationException, match="IPUs currently only support `Trainer.accumulate_grad_batches` being an int"
```
Comment (awaelchli) [inline tests/accelerators/test_ipu.py:315]:
```suggestion
        MisconfigurationException, match="IPUs currently only support `Trainer.accumulate_grad_batches` being an int."
```
don't forget the period :)

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 6]
```diff
@@ -185,34 +186,26 @@ def _convert_to_poptorch_loader(
...
         """
-        if self.accumulate_grad_batches > 1:
+        accumulate_grad_batches = self.lightning_module.trainer.accumulate_grad_batches
+        if not isinstance(accumulate_grad_batches, int):
+            raise MisconfigurationException(
+                "IPUs currently only support `Trainer.accumulate_grad_batches` being an integer."
+                f" Received {accumulate_grad_batches}"
+            )
+        # save the original value which will be used to update the global step progress
+        self._original_accumulate_grad_batches = accumulate_grad_batches
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ipu.py:204]: you are still using this attribute here, so it should not be removed in init

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 7]
```diff
@@ -96,7 +96,6 @@ def __init__(
         self.autoreport = autoreport
         self.autoreport_dir = autoreport_dir
         self.poptorch_models = {}
-        self._original_accumulate_grad_batches = None
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/ipu.py:99]: below it is still used

**Review by awaelchli — APPROVED** [review]

File: tests/accelerators/test_ipu.py (hunk 5 above)
Comment (carmocca) [reply tests/accelerators/test_ipu.py:315]: `pytest.raises` doesn't require the full match, matching a subset works. I prefer matching against the minimum possible message that clearly shows what's intended to test so that any future minor changes don't necessarily fail the test.

And if we wanted to test the grammar itself, it'd be `\.`, otherwise it'd match against single character as it gets interpreted as regex by default.

File: tests/plugins/test_deepspeed_plugin.py (hunk 3 above)
Comment (carmocca) [reply tests/plugins/test_deepspeed_plugin.py:722]: The problem is we get

```python
accumulate_grad_batches = 2

epoch, batch_idx, our_global_step, deepspeed_global_step
0, 0, 0, 0
0, 1, 0, 0 <- both step
0, 2, 1, 1
0, 3, 1, 1 <- both step
0, 4, 2, 2 <- is_last_batch=True, lightning normally steps but since deepspeed is used, it doesn't

1, 5, 2, 2 <- deepspeed steps because it doesnt reset the accumulation at the end of epoch
1, 6, 2, 3 <- lightning steps
1, 7, 3, 3
```

As you can see, Lightning and DeepSpeed go out-of-sync

Comment (tchaton) [reply tests/plugins/test_deepspeed_plugin.py:722]: Yes, I know. This is expected :)

**Review by kaushikb11 — APPROVED** [review]
